File: sdk/flutter/android/src/main/kotlin/com/openstory/sdk/internal/cache/ViewedStorySession.kt
package com.openstory.sdk.internal.cache

import com.openstory.sdk.internal.network.SdkFeedGroupPayload

internal class ViewedStorySession(
    initialSnapshot: ViewedStoryStateSnapshot,
    private val onStoryViewed: (String) -> Unit,
) {
    private var snapshot = initialSnapshot

    fun isGroupViewed(group: SdkFeedGroupPayload): Boolean {
        return snapshot.isGroupViewed(group)
    }

    fun firstUnviewedStoryIndex(group: SdkFeedGroupPayload): Int {
        return snapshot.firstUnviewedStoryIndex(group)
    }

    fun markStoryViewed(storyRevisionId: String) {
        val nextSnapshot = snapshot.withViewedStory(storyRevisionId)
        if (nextSnapshot == snapshot) {
            return
        }

        snapshot = nextSnapshot
        onStoryViewed(storyRevisionId)
    }
}

File: sdk/flutter/ios/open_story_flutter/Sources/open_story_flutter/OpenStorySDK/Internal/Cache/ViewedStorySession.swift
import Foundation

internal final class ViewedStorySession {
    private var snapshot: ViewedStoryStateSnapshot
    private let onStoryViewed: (String) -> Void

    init(
        initialSnapshot: ViewedStoryStateSnapshot,
        onStoryViewed: @escaping (String) -> Void
    ) {
        snapshot = initialSnapshot
        self.onStoryViewed = onStoryViewed
    }

    func isGroupViewed(_ group: SdkFeedGroupPayload) -> Bool {
        snapshot.isGroupViewed(group)
    }

    func firstUnviewedStoryIndex(in group: SdkFeedGroupPayload) -> Int {
        snapshot.firstUnviewedStoryIndex(in: group)
    }

    func markStoryViewed(_ storyRevisionId: String) {
        let nextSnapshot = snapshot.withViewedStory(storyRevisionId)
        guard nextSnapshot != snapshot else {
            return
        }

        snapshot = nextSnapshot
        onStoryViewed(storyRevisionId)
    }
}
